tensorflow_model_optimization/python/core/sparsity_tf2/sparse_utils.py

```diff
@@ -0,0 +1,71 @@
+class Bernouilli(tf.keras.initializers.Initializer):
+  """
+  Initialization distributio following a Bernouilli process..
+  """
+
+  def __init__(self, p):
+    """
+    p: probability parameter of success (i.e. 1).
+    """
+    self.p = p
+
+  def get_config(self):
+    return {'p': self.p}
+
+  @classmethod
+  def from_config(cls, config):
+    return cls(**config)
+
+  def __call__(self, shape, dytpe=tf.dtypes.float32):
+    """Number of zeros = np.ceil(sparsity * size) in expectation."""
+    probs = tf.zeros(shape=list(shape)) + self.p
+    uniform = tf.random.uniform(shape)
+    initial = tf.less(uniform, probs)
+
+    return tf.cast(initial, dtype=dtype)
+
+class PermuteOnes(tf.keras.initializers.Initializer):
+  """
+  Initialization of a deterministically sparse matrix.
+  """
+  def __init__(self, p=None):
+    """
+    p: probability parameter of success (i.e. 1).
+    If p is None, will sample randomly from uniform distribution for sparsity.
+    """
+    self.p = p if p else tf.random.uniform(())
+
+  def get_n_ones(self, shape, dtype=tf.dtypes.float32):
+    sparsity = self.p if self.p else 0.0
+    return tf.math.ceil(sparsity * tf.cast(tf.math.reduce_sum(shape)), dtype=dtype)
+
+  def __call__(self, shape, dtype=tf.dtypes.float32, seed=None):
+    flat_mask = tf.reshape(tf.ones(shape), (-1,))
+    num_ones = self.get_n_ones(shape, dtype)
+    _indices = tf.cast(tf.reshape(tf.linspace(0, num_ones - 1, int(num_ones)), (-1,)), tf.int32)
+    indices = tf.reshape(_indices, (-1, 1))
+    updates = tf.ones_like(_indices)
+    flat_shape = flat_mask.shape
+    unshuffled_mask = tf.scatter_nd(indices, udpates, flat_shape)
+    shuffled_mask = tf.random.shuffle(unshuffled_mas, seed=seed)
+
+    return tf.reshape(shuffled_mask, shape)
+
+
+class ErdosRenyi(tf.keras.Initializers.Initializer):
```

Comment:
After seeing how your code elsewhere is written (using the distributions to generate a sparsity float from the shape), I would say:

1. If you decide to pass a sparsity float to the pruner directly, your ErdosRenyi* distributions should be callables/util methods that output a sparsity float but do not extend `keras.initializers.Initializer`

2. If you decide to pass an initializer directly to the RIGLPruner *instead* of a sparsity float, it makes sense to have these two ErdosRenyi* distributions be initializers as well.

Reply:
I think I'd need to pass the initializer into the pruner because the pruner needs to create slots. The sparsity float being passed is actually the per layer sparsity, which I calculate in the config using the overall sparsity. It's a bit confusing, there's an overall target sparsity and a layer-specific sparsity (depending on the sparse distribution)

Reply:
I also cannot just grab the overall sparsity from the schedule anymore because we are no longer using a PruningSchedule, but an update Schedule object